ding/framework/supervisor.py

```diff
@@ -0,0 +1,311 @@
+import platform
+from time import sleep
+import traceback
+import uuid
+from ditk import logging
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, List, Optional, Union
+from enum import Enum
+
+
+@dataclass
+class SendPayload:
+    proc_id: int
+    req_id: str = field(default_factory=lambda: uuid.uuid1().hex)
+    method: str = None
+    args: List = field(default_factory=list)
+    kwargs: Dict = field(default_factory=dict)
+
+
+@dataclass
+class RecvPayload:
+    proc_id: int
+    req_id: str = None
+    method: str = None
+    data: Any = None
+    err: Exception = None
+
+
+class ReserveMethod(Enum):
+    SHUTDOWN = "_shutdown"
+    GETATTR = "_getattr"
+
+
+class ChildType(Enum):
+    PROCESS = "process"
+    THREAD = "thread"
+
+
+class Child:
+    """
+    Abstract class of child process/thread.
+    """
+
+    def __init__(self, proc_id: int, init: Callable, args: List[Any], recv_queue: Union[mp.Queue, queue.Queue]) -> None:
+        self._proc_id = proc_id
+        self._init = init
+        self._args = args
+        self._recv_queue = recv_queue
+
+    def start(self):
+        raise NotImplementedError
+
+    def restart(self):
+        raise NotImplementedError
+
+    def shutdown(self, timeout: Optional[float] = None):
+        raise NotImplementedError
+
+    def send(self, payload: SendPayload):
+        raise NotImplementedError
+
+    def _target(
+        self, proc_id: int, init: Callable, args: List, send_queue: Union[mp.Queue, queue.Queue],
+        recv_queue: Union[mp.Queue, queue.Queue]
+    ):
+        send_payload = SendPayload(proc_id=proc_id)
+        child_ins = init(*args)
+        while True:
+            try:
+                send_payload: SendPayload = send_queue.get()
+                if send_payload.method == ReserveMethod.SHUTDOWN:
+                    break
+                if send_payload.method == ReserveMethod.GETATTR:
+                    data = getattr(child_ins, send_payload.args[0])
+                else:
+                    data = getattr(child_ins, send_payload.method)(*send_payload.args, **send_payload.kwargs)
+                recv_queue.put(
+                    RecvPayload(proc_id=proc_id, req_id=send_payload.req_id, method=send_payload.method, data=data)
+                )
+            except Exception as e:
+                logging.warning(traceback.format_exc())
+                logging.warning("Error in child process! id: {}, error: {}".format(self._proc_id, e))
+                recv_payload = RecvPayload(
+                    proc_id=proc_id, req_id=send_payload.req_id, method=send_payload.method, err=e
+                )
+                recv_queue.put(recv_payload)
+
+    def __del__(self):
+        self.shutdown()
+
+
+class ChildProcess(Child):
+
+    def __init__(self, proc_id: int, init: Callable, args: List[Any], recv_queue: mp.Queue) -> None:
+        super().__init__(proc_id, init, args, recv_queue)
+        self._send_queue = mp.Queue()
```

Comment:
can we use [plasma](https://arrow.apache.org/docs/python/plasma.html) to replace queue for IPC?

Reply:
I don't know much about plasma, but from the documentation it looks like it's positioned differently from queue. To keep things simple, let's use queue for now, and see what plasma does in the future when we get a chance.